### PR TITLE
WIP: fix invalid check for is_inherent_required

### PR DESCRIPTION
**Pull request by thiolliere**
Related https://github.com/paritytech/substrate/issues/6243

Make pallet which provide an inherent required to provide one and create a new hook `check_inherent_data` for pallet which just want to check inherent data.

### CODE REVIEW

**Comment by bkchr** [comment]
Not sure if removing it, is the same as fixing the problem :P

File: frame/support/src/dispatch.rs [hunk 1]
```diff
@@ -306,6 +306,9 @@ impl<T> Parameter for T where T: Codec + EncodeLike + Clone + Eq + fmt::Debug {}
 /// * `integrity_test`: Executes in a test generated by `construct_runtime`, note it doesn't
 ///   execute in an externalities-provided environment. Implement
 ///   [`IntegrityTest`](./trait.IntegrityTest.html) trait.
+/// * `check_inherent_data`: check block in regards to inherent data provided. This is typically
```
Comment (bkchr) [inline frame/support/src/dispatch.rs:309]: Why do you want to move this here?

Could you explain what your idea is here for the implementation?

**Review by bkchr — COMMENTED** [review]
Would be nice if you explain what you now want to implement here.

**Comment by thiolliere** [comment]
> Would be nice if you explain what you now want to implement here.

I'm not very sure of the direction yet but here is my current thought:

* If `ProvideInherent::create_inherent` returns an option then it is hard to check that other pallet inherent are included or not. Like if pallet 5 return `None` and pallet 6 return `Some(..)`, how can we know which inherent correspond to which pallet. So we can still do that `ProvideInherent::check_inherent` return whereas it skipped the inherent, or we can make inherent always required. (I'm not very fixed yet, I find the former easier, but making inherent always required can be annoying too).
  So currently the PR makes `create_inherent` returning the required call.

* Some pallet such as babe doesn't provide any inherent, but rather just want to assert more constraint on the timestamp inherent. I think we can generalize this: the pallet can just check the calls and storage in regards to some inherent data.
  Thus I introduced the new hook: `check_inherent_data` in decl_module which have inherent data in argument, and also maybe the block extrinsics.

also I think this needs some few changes needed on ProvideInherentResult.

**Comment by thiolliere** [comment]
all in all I think this PR goes into the wrong direction, instead I'll open a new PR which makes clear what `is_inherent_required` checks for with a warning.
And I'll warn also about not being confused with ValidateUnsigned.

**Pull request closed without merging**